### DISCUSSION TITLE
Feature request: add more QuickNav command in browse mode

### DISCUSSION THREAD

**Discussion by mltony** [nvaccess/nvda#15996] [Ideas]
### Is your feature request related to a problem? Please describe.
I propose to add more QuickNav commands to browse mode to allow new types of navigation on web pages. I I will implement this feature request myself if green light is given by NVDA team. Here are new QuickNav commands I would like to implement:
#### Jump  to next/previous text paragraph
* Here we define text paragraph as paragraph with at least one sentence, that is written with punctuation marks, such as period, question mark, exclamation mark, their full-width CJK counterparts, etc.
* This allows users to quickly skip over clutter (such as menus, ads, etc) on a wide variety of websites.
* Another illustration of its utility is finding the beginning of an article on an unfamiliar news web site.
* This function is a heuristic that doesn't work 100% of the time. There are ads written in complete sentences, and likewise, sometimes in readable text of an article there can be a paragraph without any punctuation marks that will be missed. However, in practice this function works very well most of the time on a wide range of websites and simplifies navigation significantly.
* This feature has already been implemented in [TextNav add-on](https://addons.nvda-project.org/addons/textnav.en.html) and by default assigned to `Alt+Shift+Up/Down` gestures.
* The add-on is quite stable, it has been around for 5 years and for the last 4 years it's been in maintenance mode and didn't require any fixing besides NVDA compatibility updates.
* TextNav add-on is quite popular among NVDA users and has been frequently recommended on NVDA mailing list. Recently four users have conteacted me and asked to make it compatible with NVDA v2024.1. Moreover, a quick search reveals that people recommend TextNav in their blogs:
    * https://webfriendlyhelp.com/textnav/
    * (Misspelled as TestNav) https://slnconline.com.ng/10_nvda_addon_to_boost_your_productivity_and_effeciency.html
* In spirit this feature is similar to built-in N QuickNav command "Skip forward past a block of links", which in my understanding was also designed to skip menus on the websites. However in my experience jumping to text paragraph works much better then N QuickNav command as it works more reliably and typically finds exactly the first paragraph of the article.
* Finally, I believe this command will be quite useful for NVDA beginners, as one of the top tasks they struggle with is trying to make sense of web pages while trying to memorize dozens of keystrokes. With this command they'll be able to quickly and easily start using news articles, which I would think is one of the major use cases for the beginners.
* I propose to assign by default  QuickNav P and Shift+P gestures to this command.
#### Vertical navigation
* We define vertical navigation to be jumping to next or previous paragraph, which is  vertically aligned visually with the current paragraph, that is their `x` coordinate is exactly the same.
* This can be useful on websites that visually display data as a table, but for whatever reason NVDA doesn't see the table. This command can be used to navigate to next/previous cell in the same column.
* A good example of this is [issues list on github](https://github.com/nvaccess/nvda/issues/) - with this feature users will be able to find the first issue and then navigate between issues in the list using vertical navigation with a single keystroke.
* I propose to assign by default  Left bracket `[` and `Shift+[` gestures to this command.
* This feature has already been implemented in [BrowserNav add-on](https://addons.nvda-project.org/addons/browsernav.en.html) and by default assigned to `NVDA+Control+Up/Down` gestures.
#### Same style navigation
  * This provides command to find to next/previous paragraph written in the same style as current paragraph.
  * Style is defined as font name and other attributes, such as font size, bold, italic, etc.
  * For paragraph that contain text in multiple styles we will only consider the style of the first character.
  * Utility of this function can also be illustrated on [issues list on github](https://github.com/nvaccess/nvda/issues/) - as all the issue titles are written in the same style.
  * I propose to assign by default  right bracket `]` and `Shift+]` gestures to this command.
  * This feature has already been implemented in [BrowserNav add-on](https://addons.nvda-project.org/addons/browsernav.en.html) and can be triggered by first adjusting BrowserNav rotor by pressing `NVDA+O` to the value "BrowserNav navigates by  font size and same style"  and then pressing `NVDA+Control+Up/Down` keystrokes.
#### Jump to next/previous items by their roles:
* Menu item
* Toggle Button
* Progress bar
* math
* I propose to leave these gestures unassigned by default.
### Implementation details
I will implement all the features mentioned above as QuickNav commands in browse mode.
For navigating to next/previous text paratgraph, vertical navigation and style navigation:
* the implementation will be generic and will work with every browser mode implementation.
* Logic will be implemented in Python by checking all the paragraphs in the desired direction and checking one-by-one if paragraph satisfies desired criteria.
* In theory this could be optimized by moving some or all logic to C++ layer, however it appears that Python implementation (that can currently be tested in TextNav and BrowserNav add-ons) is fast enough on the vast majority of websites.
Navigating to next element by type (menu/progress bar/math) will be  only implemented for IA2, since all major browser these days work with IA2 and it's impossible to test MSHTML and UIA implementations. It will be implemented in a manner similar to other QuickNav commands.
### Personal note
I I will implement this feature request myself if green light is given by NVDA team. I am on a leave from work until early May 2024, so I will have plenty of free time until then and I promise to promptly fix any issues arising from my PRs. However after May I'll still try to find time to fix issues, but as time permits.

**Comment by cary-rowen** [comment]
cc @michaelDCurran 

Very happy to hear this and I'm really looking forward to these quick navigation features. Today's web pages are complex and diverse, and we need more solutions to flexibly access web pages.

**Comment by Brian1Gaff** [comment]
I have a couple of comments. On the whole a good idea, but there does seem if the add on web formatter is in use that nvda can stop working. Also, when using web sites for local news in the UK  punctuation is hardly ever  consistent, which makes things hard. As long  as the implementation does not cause any other interactions, it sounds like a good idea, but as always, a clear  set of instructions for what things do would be needed. Brian

-- 
***@***.*** 
Sent via blueyonder.(Virgin media) 
Please address personal E-mail to:-
***@***.***, putting 'Brian Gaff'
in the display name field.
  ----- Original Message -----

**Comment by Adriani90** [comment]
For menus / menu items there is #7695 already filled.

For paragraph style navigation, the current proposal seems to complicated in my opinion.
1. Jump to next / previous text paragraph:
* If you say the results of your add-on in the wild show that the community likes this much more than the n and shift+n command, why not extending the current behavior with part of your work? As far as I understand the n and shift+n command looks for elements with no semantic role, so you could extend that to include an additional check for punctuation and we don't need an additional key command.

2. Vertical navigation
* This is a really really useful feature. I would love to see it in NVDA core. Vertical navigation in general is commonly understood as navigation via up and down arrow keys. So I strongly suggest to maintain this logic in order to not disorientate users. I propose nvda+alt+up and down arrow (desktop) and nvda+shift+alt+up and down arrow (laptop) for this feature.

3. Same style navigation
* It is really important to have a feature like this in order to work together with sighted people and understand their navigation paterns as well. It will definitely help to understand a structure of a website in more detail.
* I would propose p and shift+p for this kind of paragraph navigation. At least form an user perspective it feels more consistent to press p and hear something like "no next paragraph of the same formating style".
* Paragraphs of the same formating style have often contextual similarities or connections, so as an user i would decide focusing the current paragraph to search for the next one with the same formating style. However, when i leave this paragraph and I land on another one with a different formating style, I might want to go back as fast as possible instead of changing the paragraph style navigation to another style unexpectedly. So i propose something like a style locking feature, like following process:
a. Focus a paragraph
b. Press nvda+shift+p --> NVDA reports: paragraph style set as reference
c. Press p or shift+p --> NVDA navigates to paragraphs of this reference style no matter where I am on the website
d. Press nvda+shift+p on another paragraph formating style --> NVDA changes the previous style reference to the current paragraph style and maintains this for the p and shift+p navigation
e. The cache for the paragraph formating style reference doesn't need to be cleaned everytime NVDA restarts or so because the design of the websites will probably not change that often. If the design changes, you can change the paragraph style reference anyway as you wish.
f. If the user didn't choose any paragraph style reference, p and shift+p command reports "no paragraph style chosen"

**Comment by XLTechie** [comment]
I would support this. Some of these are much needed features.

**Comment by mltony** [comment]
@Adriani90, thanks for your comment;
Re: #7695 - it seems jump to next menu was requested there but never implemented - at least I can' find implementation in current NVDA source code.
1. Here are some reasons I don't want to update N and Shift+N commands:
    * I never figured out utility of N and Shift+N commands myself, yet there might be users who are used to current behavior. If we change that behavior, those users might complain.
    * Trying to make some sort of hybrid of my proposal and current N command will likely make text paragraph navigation less useful. FOr example because N is looking for text away from hyperlinks and these days news articles might contain plenty of hyperlinks, so if we keep the current logic of N command, we are likely to miss important paragraphs.
2. Vertical navigation
    * I prefer to have a bracket for vertical navigation to keep it consistent with all the other QuickNav commands - just a single key for finding next element and Shift+key for finding the previous one. However if NVDA devs insist on assigning any other shortcut by default - I am fine with that.
3. Same style navigation
    * Re keystroke: I would prefer to assign P to text paragraph navigation because text paragraph navigation is by far the most useful feature in this feature request, since it can be used on a wide variety of websites, therefore it deserves a keystroke that is easy to remember. Same style navigation on the other hand is more niche feature for more advanced users.
    * Re style locking feature, : this seems to me overcomplicated: 
        * First, when using same style navigation, we will never land on a paragraph with different style by definition. 
        * Second, the intended use case is this: on github list of issues page you find the first issue in the list via other means. Then you press `]` and `Shift+]` to navigate between issues in a single keystroke. This is supposed to be much more convenient than current way, where it takes 2-3 `Control+Down` commands to do the same. Locking paragraph style doesn't help in this use case at all.
        * In general I can see utility of locking to one specifric paragraph. But this more resembles setting a bookmark on a web page, which is already implemented in PlaceMarkers add-on and also in my BrowserNav QuickJump Bookmarks. This idea seems to be orthogonal to same style navigation. It can be proposed to add to NVDA core separately.

**Comment by mltony** [comment]
@Brian1Gaff, I didn't understande your comment about web formatter.
Re UK news websites, I tried TextNav for you with 3 random articles from UK news. On BBC and the Guardian pressing `Alt+Shift+Down` reads the first paragraph of the story, which is optimal and illustrates the utility of text paragraph navigation:
* https://www.bbc.com/news/world-africa-67826868
* https://www.theguardian.com/uk-news/2023/dec/29/tony-blair-rwanda-style-asylum-plan-2003
On local news website pressing `Alt+Shift+Down` once reads some privacy policy agreement, which happens to be written in sentences, but if you press twice - then it takes you to the beginning of the article:
* https://www.yorkpress.co.uk/news/24001621.yorks-electric-buses-just-cant-cope-winter-weather/
So from my quick investigation it seems that TextNav works on British news websites very well. If you have an example where it doesn't work well, feel free to post a link.

**Comment by seanbudd** [comment]
It would be best if each issue was separately reported, so that they can be discussed, triaged and implemented independently.

**Comment by surfer0627** [comment]
Jump to next/previous text

See issue #12506, also #10648.

**Comment by cary-rowen** [comment]
I think these features are very necessary and can be implemented step by step by breaking them into small requests.

**Comment by cary-rowen** [comment]
@mltony 
You might also want to take a look at #9527,
I really like the solution proposed by @LeonarddeR in the original description of #9527.

**Comment by mltony** [comment]
@seanbudd, Sounds good, will split into separate feature requests.
@surfer0627, one can think of many different heuristics of how to find readable content on the web page. The benefit of TextNav is that it is already battle tested and proven to be effective. For static text I am skeptical it'll work as well, since many news websites as well as wiki articles contain plenty of hyperlinks, so if I understand the term static text correctly to be text without any formatting including hyperlinks, then this approach might skip many essential paragraphs. 
@cary-rowen, I already have navigate to next style change in BrowserNav bound to `~` key. I haven't found it very useful in my own experience, so I am not very enthusiastic about it.

**Reply by surfer0627** [reply]
> so if I understand the term static text correctly to be text without any formatting including hyperlinks, then this approach might skip many essential paragraphs.

Ah, Thank you @mltony for this point.

I suggest, perhaps, @Adriani90 could consider reopen issue 10648. It is more related to the request "Jump to next/previous text".

Previously, the connection between issue 12506 and the feature "Jump to next/previous text" was established for a reason. 
It was based on the following message: "Closing in favor of 12506 which contains more discussion and a potential developer who might contribute to implement it when the expected user experience is clear enough."

**Comment by XLTechie** [comment]
> so if I understand the term static text correctly to be text without any formatting including hyperlinks, then this approach might skip many essential paragraphs.

But is that an accurate understanding?

I always understood static text to mean text that is not dynamic. That is, it 
isn't being recreated during the life of the page. Not that it didn't contain 
formatting, anchors, or other buttons or clickibles.

NVDA's "n" browse mode command, moves one to the end of a block of links. I have 
never looked at how that works, but presumably it moves out of the current 
container or block level element.

What seems most desirable here, is to move to the next container or element 
containing text. Not that the text it reaches is devoid of formatting or links. 
The fact that the web consists of hypertext, implies that it may contain links. 
Just that it isn't only made up of links, such as a menu or list would be.

**Comment by mltony** [comment]
Split this into four separate feature requests:
#15998 - add QuickNav command to jump to next/previous text paragraph
#15999 - add QuickNav command for Vertical navigation in browse mode
#16000 - add QuickNav command for Same style navigation in browse mode
#16001 - add more roles to jump to in browse mode QuickNav commands
From now on please leave your comments there.

**Comment by mltony** [comment]
@seanbudd, maybe you can lock this thread so that we keep all further discussion in new separate feature requests.
